base_layer/core/src/transactions/transaction_components/contract_definition_features.rs

```diff
@@ -0,0 +1,265 @@
+//  Copyright 2022. The Tari Project
+//
+//  Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+//  following conditions are met:
+//
+//  1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+//  disclaimer.
+//
+//  2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+//  following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+//  3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+//  products derived from this software without specific prior written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+//  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use std::io::{Error, Read, Write};
+
+use integer_encoding::VarInt;
+use serde::{Deserialize, Serialize};
+use tari_common_types::{
+    array::copy_into_fixed_array_lossy,
+    types::{FixedHash, PublicKey},
+};
+use tari_utilities::Hashable;
+
+use crate::consensus::{ConsensusDecoding, ConsensusEncoding, ConsensusEncodingSized, ConsensusHashWriter, MaxSizeVec};
+
+// Maximum number of functions allowed in a contract specification
+const MAX_FUNCTIONS: usize = u16::MAX as usize;
+
+// Fixed lenght of all string fields in the contract definition
+const STR_LEN: usize = 32;
+
+#[derive(Debug, Clone, PartialEq, Deserialize, Serialize, Eq, Hash)]
```

Comment:
```suggestion
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize, Eq)]
```
This is not the same Hash trait we use, this is the rust one `std::hash`

Reply:
I did it because `output_features.rs` (which is the "parent" struct) derives the standard that Hash trait, so it makes it required to do it here as well

Reply:
mmm, wonder why it has it...
It looks like it could be removed, but I think that's for another PR